cirq-google/cirq_google/engine/engine_sampler.py

```diff
@@ -139,3 +139,106 @@ def get_engine_sampler(
         )
     gate_set = gate_sets.NAMED_GATESETS[gate_set_name]
     return engine.get_engine(project_id).sampler(processor_id=processor_id, gate_set=gate_set)
+
+
+def get_device_sampler(
+    project_id: Optional[str] = None, processor_id: Optional[str] = None
+) -> Tuple[
+    Tuple[Union[cirq.Device], int],
+    Union['cirq_google.PhasedFSimEngineSimulator', 'cirq_google.QuantumEngineSampler'],
+    bool,
+]:
+    """Authenticates on Google Cloud, can return a Device and Simulator.
+
+    This uses the environment variable GOOGLE_CLOUD_PROJECT for the Engine
+    project_id, unless set explicitly.
+
+    Args:
+        project_id: Optional explicit Google Cloud project id. Otherwise,
+            this defaults to the environment variable GOOGLE_CLOUD_PROJECT.
+            By using an environment variable, you can avoid hard-coding
+            personal project IDs in shared code.
+        processor_id: Engine processor ID (from Cloud console or
+            ``Engine.list_processors``).
+
+    Returns:
+        A tuple of ((`Device`, `int`), `Simulator/Sampler`, `bool`). The first element is the
+        device and it's corresponding line length, the second is a simulator instance, and the
+        third is a boolean value, true if the signin was successful, false otherwise.
+    """
+    import os
+    from cirq_google import (
+        PhasedFSimEngineSimulator,
+        SQRT_ISWAP_INV_PARAMETERS,
+        PhasedFSimCharacterization,
+        Bristlecone,
+        get_engine_device,
+        get_engine_sampler,
+    )
+
+    # Converting empty strings to None for form field inputs
+    if project_id == "":
+        project_id = None
+    if processor_id == "":
+        processor_id = None
+
+    google_cloud_signin_failed: bool = False
+    if project_id is None:
+        if 'GOOGLE_CLOUD_PROJECT' not in os.environ:
+            print("No project_id provided and environment variable GOOGLE_CLOUD_PROJECT not set.")
+            google_cloud_signin_failed = True
+    else:  # pragma: no cover
+        os.environ['GOOGLE_CLOUD_PROJECT'] = project_id
+
+        def authenticate_user():
```

Comment:
It seems like this does not need to be an inline function since it is called just once immediately after being defined. Instead, I'd suggest just putting the contents of the function inline.

Reply:
Did that.